### PR TITLE
SQL Native Queue Length

### PR DESCRIPTION
**Pull request by mikeminutillo**
All the bits of #118 that ARE related to SQL

Depends on #121 

Provides an implementation of Queue Length Provider specific to the SQL transport. Accepts endpoint metadata reports and uses them to map endpoint->tables. Regularly monitors the lengths of tables and updates the Queue Length Store appropriately.

### CODE REVIEW

File: src/ServiceControl.Monitoring.SmokeTests.SQLServer/Tests/When_querying_queue_length.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            }
+
+            class Handler : IHandleMessages<SampleMessage>
+            {
+                public QueueLengthContext Context { get; set; }
+
+                public Task Handle(SampleMessage message, IMessageHandlerContext context)
+                {
+                    //Concurrency limit 1 and this should block any processing on input queue
+                    return Task.Delay(TimeSpan.FromDays(1), Context.CancelProcessingTokenSource.Token);
```
Comment (Scooletz) [inline src/ServiceControl.Monitoring.SmokeTests.SQLServer/Tests/When_querying_queue_length.cs]: That's what I call the delay ;-)

File: src/ServiceControl.Transports.SQLServer/QueueLengthProvider.cs [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+				endpointInputQueueToTableName.AddOrUpdate(endpointInputQueue, _ => sqlTable, (_, cv) =>
+				{
+					previousSqlTable = cv;
+					return sqlTable;
+				});
+
+			    if (previousSqlTable == null)
+			    {
+			        tableSizes.TryAdd(sqlTable, 0);
+                }
```
Comment (Scooletz) [inline src/ServiceControl.Transports.SQLServer/QueueLengthProvider.cs]: formatting

File: src/ServiceControl.Transports.SQLServer/QueueLengthProvider.cs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+			var localAddress = metadataReport.LocalAddress;
+			var pluginVersion = metadataReport.PluginVersion;
+
+			if (SqlTable.TryParse(localAddress, pluginVersion, out var sqlTable))
+			{
+			    SqlTable previousSqlTable = null;
+
+				endpointInputQueueToTableName.AddOrUpdate(endpointInputQueue, _ => sqlTable, (_, cv) =>
+				{
+					previousSqlTable = cv;
```
Comment (Scooletz) [inline src/ServiceControl.Transports.SQLServer/QueueLengthProvider.cs]: I think that now I understand how it works. I'd propose anonther approach. For the `(_, cv)` function, we could immiediately remove the tableSize `tableSize.TryRemove (previousSqlTable)` and just add the new value below without any condition. This, IMO, would make it easier to understand

File: src/ServiceControl.Transports.SQLServer/SqlTable.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            unchecked
+            {
+                var hashCode = (QuotedName != null ? QuotedName.GetHashCode() : 0);
+                hashCode = (hashCode * 397) ^ (QuotedSchema != null ? QuotedSchema.GetHashCode() : 0);
+                hashCode = (hashCode * 397) ^ (QuotedCatalog != null ? QuotedCatalog.GetHashCode() : 0);
+                return hashCode;
+            }
+        }
+
+        class SqlNameHelper
```
Comment (Scooletz) [inline src/ServiceControl.Transports.SQLServer/SqlTable.cs]: Any tests for this?

File: src/ServiceControl.Transports.SQLServer/QueueLengthProvider.cs [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+		{
+			var nowTicks = DateTime.UtcNow.Ticks;
+
+			foreach (var endpointToTableNamePair in endpointInputQueueToTableName)
+			{
+				store.Store(
+					new []{ new RawMessage.Entry
+					{
+						DateTicks = nowTicks,
+						Value = tableSizes.TryGetValue(endpointToTableNamePair.Value, out var size) ? size : 0
```
Comment (Scooletz) [inline src/ServiceControl.Transports.SQLServer/QueueLengthProvider.cs]: Good to see `TryGetValue` in here. We've got two concurrent dictionaries accessed concurrently.

**Comment by tmasternak** [comment]
@Scooletz I've made changes after your review.

**Comment by Scooletz** [comment]
@tmasternak @mikeminutillo Why can't I squash'n'merge for this repo?

File: src/ServiceControl.Monitoring.SmokeTests.SQLServer/Tests/When_querying_queue_length.cs (hunk 1 above)
Comment (mikeminutillo) [reply src/ServiceControl.Monitoring.SmokeTests.SQLServer/Tests/When_querying_queue_length.cs]: You can also use Timeout.Inifinite :)

**Pull request merged**